File: CodeQualityCheckerPython/фИНАЛ ПП/CodeQualityChecker/CodeQualityChecker/Services/LlmService.cs
using System.Net.Http.Headers;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

public class LlmService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _config;

    public LlmService(HttpClient httpClient, IConfiguration config)
    {
        _httpClient = httpClient;
        _config = config;
        _httpClient.BaseAddress = new Uri(_config["LlmSettings:BaseUrl"]);
        _httpClient.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue("Bearer", _config["LlmSettings:ApiKey"]);
    }

    public async Task<string> AnalyzeCodeAsync(CodeSubmission submission, AnalysisType analysisType)
    {
        var prompt = GeneratePrompt(submission.Code, submission.Language, analysisType);

        var requestData = new
        {
            model = _config["LlmSettings:ModelName"],
            messages = new[]
            {
                new { role = "system", content = GetSystemPrompt(submission.Language, analysisType) },
                new { role = "user", content = prompt }
            },
            temperature = GetTemperature(analysisType),
            max_tokens = submission.ComplexityLimit > 500 ? 4000 : 2000
        };

        string responseContent = null;

        try
        {
            var response = await _httpClient.PostAsJsonAsync("/v1/chat/completions", requestData);
            response.EnsureSuccessStatusCode();

            responseContent = await response.Content.ReadAsStringAsync();
            Console.WriteLine($"Raw response from LM Studio: {responseContent}");

            // Парсим весь ответ как JObject
            var json = JObject.Parse(responseContent);

            // Получаем содержимое сообщения
            var contentToken = json["choices"]?[0]?["message"]?["content"];

            if (contentToken == null)
            {
                return $"❌ Модель не вернула содержимое. Raw: {responseContent}";
            }

            var content = contentToken.ToString();

            // Ищем JSON-блок между ```json и ```
            var jsonStart = content.IndexOf("```json");
            var jsonEnd = content.LastIndexOf("```");

            if (jsonStart >= 0 && jsonEnd > jsonStart)
            {
                var jsonText = content.Substring(jsonStart + 7, jsonEnd - jsonStart - 7).Trim();
                return jsonText; // Возвращаем только JSON
            }

            // Если JSON-блок не найден, пробуем парсить весь ответ
            try
            {
                JToken.Parse(content); // Попытка парсинга
                return content; // Если успешно — возвращаем как есть
            }
            catch
            {
                return $"⚠️ Ответ модели не является JSON. Содержимое: {content}";
            }
        }
        catch (Exception ex)
        {
            return $"❌ Ошибка при обращении к модели: {ex.Message}. Raw: {responseContent}";
        }
    }

    private string GetSystemPrompt(string language, AnalysisType analysisType)
    {
        var basePrompt = $"Ты — эксперт по анализу кода на {language}. ";

        switch (analysisType)
        {
            case AnalysisType.SecurityOnly:
                return basePrompt + "Фокусируйся только на безопасности: SQL-инъекции, XSS, утечки данных и другие риски. Не обращай внимания на стиль кода.";

            case AnalysisType.StyleOnly:
                return basePrompt + "Фокусируйся только на стиле: соглашения об именовании, форматирование, читаемость. Не уделяй внимание функциональным ошибкам.";

            case AnalysisType.Performance:
                return basePrompt + "Фокусируйся на производительности: эффективность алгоритмов, использование памяти, оптимизация запросов. Предлагай конкретные улучшения для повышения производительности.";

            case AnalysisType.BestPractices:
                return basePrompt + "Фокусируйся на лучших практиках: принципы SOLID, паттерны проектирования, обработка ошибок, тестируемость кода.";

            case AnalysisType.CodeExamples:
                return basePrompt + "Фокусируйся на предоставлении конкретных примеров улучшенного кода. Для каждой проблемы в исходном коде — покажи, как её можно исправить.";

            default:
                return basePrompt + "Всегда отвечай в JSON-формате с детальным анализом по всем аспектам.";
        }
    }

    private string GeneratePrompt(string code, string language, AnalysisType analysisType)
    {
        var basePrompt = $@"Проанализируй следующий {language} код:
{code}

";

        switch (analysisType)
        {
            case AnalysisType.SecurityOnly:
                basePrompt += @"
Верни ответ в формате JSON с единственным свойством:
- 'security_risks': массив строк с описанием уязвимостей безопасности (SQL-инъекции, XSS, утечки данных и т.д.)

Если уязвимостей нет, верни пустой массив.
";
                break;

            case AnalysisType.StyleOnly:
                basePrompt += @"
Верни ответ в формате JSON с единственным свойством:
- 'style_issues': массив строк с нарушениями стиля кода (соглашения об именовании, форматирование, читаемость)

Если нарушений стиля нет, верни ['Стиль кода соответствует стандартам'].
";
                break;

            case AnalysisType.Performance:
                basePrompt += @"
Верни ответ в формате JSON со свойствами:
- 'performance_issues': массив строк с проблемами производительности
- 'optimization_tips': массив строк с предложениями по оптимизации
- 'complexity_score': число от 0 до 10, где 0 - очень низкая производительность, 10 - оптимальная
";
                break;

            case AnalysisType.CodeExamples:
                basePrompt += @"
Верни ответ в формате JSON со свойствами:
- 'code_suggestions': массив строк с примерами улучшенного кода в формате:
  '// Было:\\n[старый код]\\n// Стало:\\n[улучшенный код]'
- 'improvement_reasons': массив строк с объяснением, почему улучшенный код лучше
";
                break;

            case AnalysisType.BestPractices:
                basePrompt += @"
Верни ответ в формате JSON со свойствами:
- 'best_practices': массив строк с несоответствиями лучшим практикам
- 'architecture_issues': массив строк с проблемами архитектуры
- 'improvement_suggestions': массив строк с предложениями по улучшению
";
                break;

            default:
                basePrompt += @"
Верни ответ в формате JSON со свойствами:
- 'errors': массив строк с описанием ошибок,
- 'style_issues': массив строк с нарушениями стиля,
- 'security_risks': массив строк с уязвимостями,
- 'best_practices': массив строк с рекомендациями по лучшим практикам,
- 'performance_issues': массив строк с проблемами производительности,
- 'code_suggestions': массив строк с примерами улучшенного кода (если 'include_code_examples': true),
- 'summary': краткое резюме,
- 'complexity_score': число от 0 до 10 (уровень сложности/читаемости),
- 'maintainability_score': число от 0 до 10 (уровень сопровождаемости),
- 'security_score': число от 0 до 10 (уровень безопасности),
- 'performance_score': число от 0 до 10 (уровень производительности).

Пример ответа:
{
  ""errors"": [""Нет ошибок""],
  ""style_issues"": [""Можно улучшить именование переменных""],
  ""security_risks"": [""Нет уязвимостей""],
  ""best_practices"": [""Используйте константы для магических чисел""],
  ""performance_issues"": [],
  ""code_suggestions"": [
    ""// Было:\\nvar x = 42;\\n// Стало:\\nconst int magicNumber = 42;\\nvar x = magicNumber;""
  ],
  ""summary"": ""Код корректен и соответствует стилю"",
  ""complexity_score"": 7,
  ""maintainability_score"": 8,
  ""security_score"": 10,
  ""performance_score"": 9
}
";
                break;
        }

        // Добавляем контекст для конкретного языка, чтобы анализ был точнее
        switch (language)
        {
            case "Python":
                basePrompt += @"
Для Python: удели особое внимание:
- использованию `if __name__ == '__main__':`
- PEP 8 стилю (именование, отступы)
- безопасности (использование `pickle`, `eval`, `exec`)
- производительности (использование `list comprehensions`, `map`, `filter`).
";
                break;

            case "JavaScript":
                basePrompt += @"
Для JavaScript: удели особое внимание:
- безопасности (XSS, использование `eval`, `innerHTML`)
- стилю (ESLint, соглашения об именовании)
- асинхронности (`async/await`, `Promise`)
- утечкам памяти (замыкания, обработчики событий).
";
                break;

            case "Java":
                basePrompt += @"
Для Java: удели особое внимание:
- соблюдению принципов SOLID
- безопасности (SQL-инъекции, XSS)
- стилю (Google Java Style Guide)
- управлению памятью (утечки, `finalize`)
- производительности (использование `StringBuilder`, `Stream API`).
";
                break;

            case "C#":
                basePrompt += @"
Для C#: удели особое внимание:
- соблюдению .NET Framework Design Guidelines
- безопасности (XSS, SQL-инъекции, утечки памяти)
- стилю (именование, форматирование)
- использованию `async/await`, `LINQ`, `nullable reference types`.
";
                break;

            case "Go":
                basePrompt += @"
Для Go: удели особое внимание:
- соблюдению `gofmt`, `golint`, `go vet`
- безопасности (SQL-инъекции, XSS)
- производительности (использование `goroutines`, `channels`)
- стилю (идиоматичный Go).
";
                break;

            case "Rust":
                basePrompt += @"
Для Rust: удели особое внимание:
- безопасности (владение, заимствование)
- стилю (использование `clippy`, `rustfmt`)
- производительности (использование итераторов, `async`)
- использованию `unsafe`.
";
                break;

            case "PHP":
                basePrompt += @"
Для PHP: удели особое внимание:
- безопасности (SQL-инъекции, XSS, `eval`)
- стилю (PSR-12)
- производительности (кеширование, `opcache`)
- использованию `mysqli`, `PDO`.
";
                break;

            case "Ruby":
                basePrompt += @"
Для Ruby: удели особое внимание:
- стилю (Ruby Style Guide)
- безопасности (SQL-инъекции, XSS)
- использованию `Rails` best practices
- идиоматичности кода.
";
                break;

            case "Swift":
                basePrompt += @"
Для Swift: удели особое внимание:
- безопасности (ARC, утечки памяти)
- стилю (Swift API Design Guidelines)
- использованию `Optionals`, `Closures`, `Protocol-Oriented Programming`.
";
                break;

            case "Kotlin":
                basePrompt += @"
Для Kotlin: удели особое внимание:
- соблюдению Kotlin Coding Conventions
- безопасности (SQL-инъекции, XSS)
- использованию `coroutines`, `null safety`, `extension functions`.
";
                break;

            case "SQL":
                basePrompt += @"
Для SQL: удели особое внимание:
- безопасности (SQL-инъекции, использование `prepared statements`)
- производительности (индексы, `EXPLAIN`, `JOIN`)
- стилю (именование, форматирование).
";
                break;

            case "C++":
                basePrompt += @"
Для C++: удели особое внимание:
- безопасности (`buffer overflow`, `dangling pointers`, RAII)
- стилю (Google C++ Style Guide)
- использованию `STL`, `RAII`, `move semantics`, `templates`
- производительности (`memory allocation`, `cache misses`).
";
                break;

            case "TypeScript":
                basePrompt += @"
Для TypeScript: удели особое внимание:
- использованию `strict mode`, `strictNullChecks`, `strictFunctionTypes`
- типизации (`any`, `unknown`, `never`, `union types`, `generics`)
- безопасности (XSS через React bindings)
- стилю (`TSLint`, `ESLint`, `Prettier`, соглашения об именовании).
";
                break;

            case "C":
                basePrompt += @"
Для C: удели особое внимание:
- безопасности (`buffer overflow`, `unsafe functions`, `scanf`, `gets`)
- стилю (именование, форматирование)
- управлению памятью (утечки, `malloc/free`)
- использованию `const`, `static`, `restrict`.
";
                break;

            case "Objective-C":
                basePrompt += @"
Для Objective-C: удели особое внимание:
- безопасности (ARC, утечки памяти)
- стилю (Apple Coding Guidelines)
- использованию `Blocks`, `Categories`, `Protocols`.
";
                break;
        }

        return basePrompt;
    }

    private double GetTemperature(AnalysisType analysisType)
    {
        switch (analysisType)
        {
            case AnalysisType.SecurityOnly:
                return 0.1; // Минимальная креативность для безопасности
            case AnalysisType.Performance:
                return 0.3; // Средняя креативность для предложений по оптимизации
            case AnalysisType.CodeExamples:
                return 0.4; // Более высокая креативность для генерации примеров
            default:
                return 0.2; // Стандартная температура
        }
    }
}

File: CodeQualityCheckerPython/services/llm_service.py
"""
Сервис для работы с LLM (LM Studio / OpenAI-совместимый API).
"""
import json

import requests

from models import AnalysisType, CodeSubmission


class LlmService:
    def __init__(
        self,
        base_url: str = "http://127.0.0.1:1234",
        model_name: str = "deepseek-r1-distill-qwen-14b",
        api_key: str = "lm-studio",
    ):
        self.base_url = base_url.rstrip("/")
        self.model_name = model_name
        self.api_key = api_key
        self.session = requests.Session()
        self.session.headers.update(
            {
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            }
        )

    async def analyze_code_async(self, submission: CodeSubmission) -> str:
        model_ready, model_message = self._ensure_model_ready()
        if not model_ready:
            return json.dumps({"error": model_message}, ensure_ascii=False, indent=2)

        original_len = len(submission.code or "")
        safe_code = self._fit_code_to_context(submission.code or "")
        prompt = self._generate_prompt(
            safe_code, submission.language, submission.analysis_type
        )

        # If we had to truncate input, ask the model to explicitly mention it.
        if len(safe_code) < original_len:
            prompt += (
                "\n\nВажно: код был автоматически сокращен из-за лимита контекста модели. "
                "Укажи в summary, что анализ частичный."
            )

        completion_budget = 1200 if len(safe_code) > 6000 else 1800
        request_data = {
            "model": self.model_name,
            "messages": [
                {
                    "role": "system",
                    "content": self._get_system_prompt(
                        submission.language, submission.analysis_type
                    ),
                },
                {"role": "user", "content": prompt},
            ],
            "temperature": self._get_temperature(submission.analysis_type),
            "max_tokens": completion_budget,
        }
        try:
            response = self.session.post(
                f"{self.base_url}/v1/chat/completions",
                json=request_data,
                timeout=1200,
            )
            response.raise_for_status()
            result = response.json()
            content = result.get("choices", [{}])[0].get("message", {}).get("content", "")
            return self._extract_json(content)
        except requests.HTTPError as exc:
            details = ""
            if exc.response is not None:
                try:
                    details = exc.response.text[:1000]
                except Exception:
                    details = str(exc)
            return json.dumps(
                {"error": f"Ошибка HTTP от модели: {exc}", "details": details},
                ensure_ascii=False,
                indent=2,
            )
        except requests.RequestException as exc:
            return json.dumps(
                {"error": f"Ошибка при обращении к модели: {exc}"},
                ensure_ascii=False,
                indent=2,
            )

    def _ensure_model_ready(self) -> tuple[bool, str]:
        try:
            response = self.session.get(f"{self.base_url}/v1/models", timeout=8)
            response.raise_for_status()
            data = response.json()
            models = data.get("data", [])
            model_ids = [item.get("id", "") for item in models if isinstance(item, dict)]
            if not model_ids:
                return (
                    False,
                    "LM Studio доступен, но нет загруженных моделей. Дождитесь загрузки модели до 100%.",
                )
            if self.model_name not in model_ids:
                self.model_name = model_ids[0]
            return True, "ok"
        except requests.RequestException as exc:
            return (
                False,
                (
                    "Нет подключения к LM Studio (http://127.0.0.1:1234). "
                    f"Проверьте статус сервера в LM Studio. Детали: {exc}"
                ),
            )

    def _extract_json(self, content: str) -> str:
        json_start = content.find("```json")
        json_end = content.rfind("```")
        if json_start >= 0 and json_end > json_start:
            candidate = content[json_start + 7 : json_end].strip()
            try:
                json.loads(candidate)
                return candidate
            except json.JSONDecodeError:
                pass
        try:
            json.loads(content)
            return content
        except json.JSONDecodeError:
            return json.dumps(
                {"error": "Ответ не является JSON", "raw": content},
                ensure_ascii=False,
                indent=2,
            )

    def _get_system_prompt(self, language: str, analysis_type: str) -> str:
        base = f"Ты эксперт по анализу кода на {language}. "
        type_map = {
            AnalysisType.SECURITY_ONLY: base + "Фокусируйся только на безопасности.",
            AnalysisType.STYLE_ONLY: base + "Фокусируйся только на стиле кода.",
            AnalysisType.PERFORMANCE: base + "Фокусируйся на производительности.",
            AnalysisType.BEST_PRACTICES: base + "Фокусируйся на лучших практиках.",
            AnalysisType.CODE_EXAMPLES: base + "Фокусируйся на примерах улучшенного кода.",
        }
        return type_map.get(
            analysis_type, base + "Всегда отвечай строго валидным JSON."
        )

    def _generate_prompt(self, code: str, language: str, analysis_type: str) -> str:
        prompt = f"Проанализируй следующий {language} код:\n```{language}\n{code}\n```\n\n"
        if analysis_type == AnalysisType.SECURITY_ONLY:
            prompt += "Верни JSON с полем security_risks (массив)."
        elif analysis_type == AnalysisType.STYLE_ONLY:
            prompt += "Верни JSON с полем style_issues (массив)."
        elif analysis_type == AnalysisType.PERFORMANCE:
            prompt += "Верни JSON с полями performance_issues и optimization_tips."
        elif analysis_type == AnalysisType.CODE_EXAMPLES:
            prompt += "Верни JSON с полем code_suggestions (примеры улучшенного кода)."
        elif analysis_type == AnalysisType.BEST_PRACTICES:
            prompt += "Верни JSON с полями best_practices и architecture_issues."
        else:
            prompt += (
                "Верни JSON с полями errors, style_issues, security_risks, best_practices, "
                "performance_issues, code_suggestions, summary, complexity_score, "
                "maintainability_score, security_score, performance_score."
            )
        lang_specific = {
            "Python": "Учитывай PEP8, безопасность eval/exec/pickle, типизацию.",
            "JavaScript": "Учитывай XSS, async/await, ESLint, утечки памяти.",
            "Java": "Учитывай SOLID, потоки, безопасность, коллекции.",
            "C#": "Учитывай .NET guidelines, nullable, async/await.",
            "Go": "Учитывай gofmt, goroutines, race conditions.",
            "Rust": "Учитывай ownership, borrow checker, unsafe.",
        }
        if language in lang_specific:
            prompt += f"\n\n{lang_specific[language]}"
        return prompt

    def _get_temperature(self, analysis_type: str) -> float:
        temp_map = {
            AnalysisType.SECURITY_ONLY: 0.1,
            AnalysisType.PERFORMANCE: 0.3,
            AnalysisType.CODE_EXAMPLES: 0.4,
        }
        return temp_map.get(analysis_type, 0.2)

    def _fit_code_to_context(self, code: str) -> str:
        # Conservative cap to stay below 4k context for small local models.
        # Approximation: 1 token ~= 3-4 chars for mixed source code.
        max_chars = 7500
        if len(code) <= max_chars:
            return code

        head = code[:3500]
        tail = code[-3500:]
        return (
            f"{head}\n\n"
            "... [TRUNCATED FOR CONTEXT LIMIT] ...\n\n"
            f"{tail}"
        )
